src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -222,6 +226,33 @@ public bool RecoverDatabase(TransactionHeader* txHeader, Action<string> addToIni
                 }
             }
 
+            if (_env.Options.EncryptionEnabled == false) // for encryption, we already use AEAD, so no need
+            {
+                // here we want to check that the checksum on all the modified pages is valid
+                // we can't do that during the journal application process because we may have modifications
+                // to pages that overwrite one another. So we have to do this at the end, this will detect
+                // corruption when applying journals at recovery time rather than at usage.
+                var tempTx = new TempPagerTransaction();
+
+                var sortedPages = modifiedPages.ToArray();
+                Array.Sort(sortedPages);
+
+                long minPageChecked = -1;
+
+                foreach (var modifedPage in modifiedPages)
+                {
+                    if (minPageChecked >= modifedPage)
+                        continue;
+
+                    var ptr = (PageHeader*)_dataPager.AcquirePagePointerWithOverflowHandling(tempTx, modifedPage, null);
+                    _env.ValidatePageChecksum(modifedPage, ptr);
+
+                    minPageChecked = modifedPage + VirtualPagerLegacyExtensions.GetNumberOfPages(ptr);
```

Comment:
BTW we could unintentionally skip validation of some pages as this line should be:
`modifedPage + VirtualPagerLegacyExtensions.GetNumberOfPages(ptr) - 1` (note `- 1`)